romankondratev-BJS2-61247

romankondratev opened the pull request:
Распределенная армия героев меча и магии

romankondratev committed:
BJS2-61247 feat: implement army composition and power calculation

- Added `Unit` abstract class with power attribute.
- Implemented character classes: `Archer`, `Mage`, and `Swordsman`.
- Created `Squad` to group units and calculate their combined power.
- Developed `Army` to manage squads and calculate total army power.
- Added `Main` for demonstration and logging of army functionality.

romankondratev committed:
BJS2-61247 refactor: move character classes to 'unit' package

- Relocated `Unit`, `Archer`, `Mage`, and `Swordsman` classes from 'character' to 'unit' package.
- Updated imports across all affected files to reflect package change.

RuslanZinovyev commented on src/main/java/school/faang/task_61247/unit/Swordsman.java:
```diff
@@ -0,0 +1,9 @@
+    private static final int SWORDSMAN_POWER = 75;
+
+    public Swordsman() {
+        super(SWORDSMAN_POWER);
```
плохая практика, фактически хардкод, если я захочу параметр завтра изменить мне что надо в класс идти изменения делать? 
```
public Swordsman(int power) {
        super(power);
    }
```

RuslanZinovyev commented on src/main/java/school/faang/task_61247/unit/Mage.java:
```diff
@@ -0,0 +1,9 @@
+    private static final int MAGE_POWER = 50;
+
+    public Mage() {
+        super(MAGE_POWER);
```
```
public Mage(int power) {
        super(power);
      }
```

RuslanZinovyev commented on src/main/java/school/faang/task_61247/unit/Archer.java:
```diff
@@ -0,0 +1,9 @@
+    private static final int ARCHER_POWER = 25;
+
+    public Archer() {
+        super(ARCHER_POWER);
```
```
public Archer(int power) {
        super(power);
    }
```

RuslanZinovyev commented on src/main/java/school/faang/task_61247/Army.java:
```diff
@@ -0,0 +1,35 @@
+        final List<Integer> results = Collections.synchronizedList(new ArrayList<>());
+
+        for (Squad<?> squad : squads) {
+            final Thread thread = new Thread(() -> results.add(squad.calculateSquadPower()));
```
тут не нужно явно писать final

romankondratev committed:
BJS2-61247 refactor: добавить настройку мощности юнитов и оптимизировать код

- Убраны жестко заданные значения мощности юнитов; добавлены параметризированные конструкторы для `Mage`, `Archer`, `Swordsman`
- Добавлены константы мощности по умолчанию в `Main` для упрощения упорядоченности и гибкости
- Оптимизированы операции со стримами в `Army` для улучшения читаемости кода
- Удалены неиспользуемые статические константы в классах юнитов
